python/cudf/cudf/core/buffer.py

```diff
@@ -1,218 +1,61 @@
 
-from cudf.utils import cudautils, utils
 
-
-class Buffer(object):
-    """A 1D gpu buffer.
-    """
-
-    _cached_ipch = None
-
-    @classmethod
-    def from_empty(cls, mem, size=0):
-        """From empty device array
-        """
-        return cls(mem, size=size, capacity=mem.size)
-
-    @classmethod
-    def null(cls, dtype):
-        """Create a "null" buffer with a zero-sized device array.
-        """
-        mem = rmm.device_array(0, dtype=dtype)
-        return cls(mem, size=0, capacity=0)
-
-    def __init__(self, mem, size=None, capacity=None, categorical=False):
-        if size is None:
-            if categorical:
-                size = len(mem)
-            elif hasattr(mem, "__len__"):
-                if hasattr(mem, "ndim") and mem.ndim == 0:
-                    pass
-                elif len(mem) == 0:
-                    size = 0
-            if hasattr(mem, "size"):
-                size = mem.size
-        if capacity is None:
-            capacity = size
-        self.mem = cudautils.to_device(mem)
-        _BufferSentry(self.mem).ndim(1)
-        self.size = size
-        self.capacity = capacity
-        self.dtype = self.mem.dtype
-
-    def serialize(self):
-        """Called when dask.distributed is performing a serialization on this
-        object.
-
-        Do not use this directly.  It is invoked by dask.distributed.
-
-        Parameters
-        ----------
-        serialize : callable
-             Used to serialize data that needs serialization .
-        context : dict; optional
-            If not ``None``, it contains information about the destination.
-
-        Returns
-        -------
-        (header, frames)
-            See custom serialization documentation in dask.distributed.
-        """
-        header = {}
-        header["type"] = pickle.dumps(type(self))
-        header["cuda_array_interface"] = self.mem.__cuda_array_interface__
-        return header, [self.mem]
-
-    @classmethod
-    def deserialize(cls, header, frames):
-        """Called when dask.distributed is performing a deserialization for
-        data of this class.
-
-        Do not use this directly.  It is invoked by dask.distributed.
-
-        Parameters
-        ----------
-
-        deserialize : callable
-             Used to deserialize data that needs further deserialization .
-        header, frames : dict
-            See custom serialization documentation in dask.distributed.
-
-        Returns
-        -------
-        obj : Buffer
-            Returns an instance of Buffer.
-        """
-        assert len(frames) == 1, "Use Buffer.serialize() for serialization"
-        iface = header["cuda_array_interface"]
-        if (
-            len(iface["shape"]) == 0 or 0 in iface["shape"]
-        ):  # The array is empty
-            arr = rmm.device_array(
-                iface["shape"], dtype=np.dtype(iface["typestr"])
-            )
+class Buffer:
+    def __init__(self, ptr, size=None, owner=None):
```

Comment:
What are your thoughts on redefining this as:
```
def __init__(self, obj=None, size=None):
```
That way you can just do `Buffer(np.array(...))` or `Buffer(cupy.array(...))` or `Buffer(numba.devicearray(...))` instead of having to call `Buffer.from_array_like(...)`

If a user just does `Buffer()` it returns an instance with `self.ptr = 0` and `self.size = 0`, a user can still allocate memory themselves via `Buffer(size=10)`. This way `self.ptr` is an implementation detail within the class instead of part of the API that can allow doing some unsafe things.